### PR TITLE
Draft for Min/Max intrinsics xarch (#65625)

### PR DESCRIPTION
**Pull request by SkiFoD**
I made the changes using [#65584](https://github.com/dotnet/runtime/pull/65584) as a referrence. It works with HelloWorld'sh code, but I'm stuck with building it via build.cmd. It doesn't want to compile and I can't figure out why. 
I face the errors:
```
Assertion failed 'IsThreeOperandAVXInstruction(ins)' in 'System.MathF:Min(float,float):float' during 'Generate code' (IL size 13)
Assertion failed 'IsThreeOperandAVXInstruction(ins)' in 'System.Single:System.INumber<System.Single>.Min(float,float):float' during 'Generate code' (IL size 8)
```
image
@EgorBo Could you please help me to figure out what I implemented wrong?

### CODE REVIEW

**Comment by EgorBo** [comment]
> @EgorBo Could you please help me to figure out what I implemented wrong?

maxss and vmaxss are different instructions basically (VEX encoding), only vmaxss can be used with R_R_R, but it's not available without AVX

File: src/coreclr/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -7400,6 +7400,20 @@ void CodeGen::genIntrinsic(GenTree* treeNode)
     // Handle intrinsics that can be implemented by target-specific instructions
     switch (treeNode->AsIntrinsic()->gtIntrinsicName)
     {
+#ifdef TARGET_XARCH
+
+        case NI_System_Math_Max:
+            genConsumeOperands(treeNode->AsOp());
+            GetEmitter()->emitIns_R_R_R(INS_maxss, emitActualTypeSize(treeNode), treeNode->GetRegNum(),
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/jit/codegenxarch.cpp:7407]: I'd actually also expect we want to handle things like containment/etc in which case lowering to a HWIntrinsic might be simpler overall

[the remark this answers is not in the document]
Comment (SkiFoD) [reply src/coreclr/jit/codegenxarch.cpp:7407]: @tannergooding Thank you for the advice. I'm sorry , but could you tell me more about the containments?

**Comment by EgorBo** [comment]
I assume failures are related e.g.:
```
Process terminated. Assertion failed.
   at System.Globalization.CalendricalCalculationsHelper.PersianNewYearOnOrBefore(Int64 numberOfDays) in /_/src/libraries/System.Private.CoreLib/src/System/Globalization/CalendricalCalculationsHelper.cs:line 396
```
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-65700-merge-f505ab65165146859e/System.Runtime.Tests/1/console.e40913b6.log?sv=2019-07-07&se=2022-03-21T10%3A27%3A05Z&sr=c&sp=rl&sig=8MkYay3CV4QZ0iVUwk2q8n54366%2F0ilLMZ0UWu%2FbUms%3D

**Comment by SkiFoD** [comment]
@EgorBo Since I'm learning to work in the project, I got a question. Let's say you have a piece of code failing (like the assertion in the log). How do you run clr against this piece of code to debug what is wrong? I believe there must be a better way than just "copy pasting" the code into a simple "Hello world" project to emulate the behaviour.  Thank you in advance.

**Comment by tannergooding** [comment]
> The second questions is, I can't figure out which ISA to use for converting NI_System_Math_Max to a relevant HW node. Could you give me a hand please?

Sorry, I had missed this. You can generally see the two approaches if you look at `NI_System_Math_Round` and `NI_System_Math_FusedMultiplyAdd`. You'll see that `Compiler::IsTargetIntrinsic` reports if the given math call can be handled via special logic here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/importer.cpp#L20682-L20745

`Compiler::impIntrinsic` then specially handles importing `Math_FusedMultiplyAdd` directly as the relevant nodes here: https://github.com/dotnet/runtime/blob/b5381bf99531b88a7ddaad7a03ccacf1b92fff63/src/coreclr/jit/importer.cpp#L4390-L4459. It does this because `FusedMultiplyAdd` is a slightly more complex sequence of nodes.

Alternatively, many of the other `Math_*` intrinsics defer to `impMathIntrinsic` here: https://github.com/dotnet/runtime/blob/b5381bf99531b88a7ddaad7a03ccacf1b92fff63/src/coreclr/jit/importer.cpp#L4461-L4498. Those get handling in a few phases throughout the JIT including containment handling here: https://github.com/dotnet/runtime/blob/b5381bf99531b88a7ddaad7a03ccacf1b92fff63/src/coreclr/jit/lowerxarch.cpp#L5323-L5346. And then code generation here: https://github.com/dotnet/runtime/blob/b5381bf99531b88a7ddaad7a03ccacf1b92fff63/src/coreclr/jit/codegenxarch.cpp#L7421-L7457

**Comment by SkiFoD** [comment]
@EgorBo @tannergooding I'm sorry, I had been busy recently and the PR was closed. Can I open it again?

**Comment by tannergooding** [comment]
Yep, feel free to reopen it whenever you have time again

**Pull request closed without merging**